odoo/models.py

```diff
@@ -1731,19 +1768,326 @@ def clear_caches(cls):
+                orderby_terms.append(f'{sql_expression} {order_orientation}')
+
+        return orderby_terms, extra_groupby_terms, fnames_used
+
+    @api.model
+    def _read_group_empty_value(self, spec):
+        if spec == '__count':
+            return 0
+        fname, func = split_read_group_spec(spec)  # func is either None, granularity or an aggregate
+        if func in ('count', 'count_distinct'):
+            return 0
+        if func in ('array_agg', 'array_agg_distinct'):
+            return []
+        field = self._fields[fname]
+        if not func and (field.relational or fname == 'id'):
+            return self.env[field.comodel_name] if field.relational else self.env[self._name]
+        return False
+
+    @api.model
+    def _read_group_postprocess(self, spec, raw_values):
+        """ Return format function the specific _read_group ``spec`` (groupby and aggregate).
+
+        This format is for the backend point of view. Rules are:
+        - ``spec`` related to a Relational field is converted to recordset with correct prefetch
+        - ``spec`` related to a Datetime field is localized with the context tz
+        - Other ``spec`` is only change None value by False
+        """
+        def get_convert_method():
+            empty_value = self._read_group_empty_value(spec)
+            if spec == '__count':
+                return lambda value: value if value is not None else empty_value
+
+            fname, func = split_read_group_spec(spec)
+            field = self._fields[fname]
+
+            if field.type == 'datetime' and self._context.get('tz') in pytz.all_timezones_set:
+                tz = pytz.timezone(self._context['tz'])
+                if func and func in ('array_agg', 'array_agg_distinct'):
+                    return lambda values: [tz.localize(value) for value in values if value] if values else empty_value
+                return lambda value: tz.localize(value) if value is not None else empty_value
+
+            if func and func in ('array_agg', 'array_agg_distinct'):
+                return lambda value: value if value is not None else empty_value
+
+            if not func and (field.relational or fname == 'id'):
+                Model = self.pool[field.comodel_name] if field.relational else self.pool[self._name]
+                prefetch_ids = tuple(raw_value for raw_value in raw_values if raw_value)
+                browse_partial = functools.partial(Model, env=self.env, prefetch_ids=prefetch_ids)
+                return lambda value: browse_partial(ids=(value,)) if value else empty_value
+
+            return lambda value: value if value is not None else empty_value
+
+        return map(get_convert_method(), raw_values)
+
+    @api.model
+    def _read_group_check_field_access_rights(self, field_names):
+        # TODO: should have a general hook (not only for read_group) for this kind of stuff
+        self.check_field_access_rights('read', field_names)
```

Comment:
Yep, but I cannot restrict in `check_field_access_rights` (or check with call stack whatever I came from `_read_group`) because it is used by `_read` (and we don't want restrict for these cases).